main.nf

```diff
@@ -972,66 +979,157 @@ process dupradar {
+              else if (filename.indexOf("_gene.htseq-count.txt") > 0) "gene_counts/$filename"
+              else "$filename"
+          }
 
-    input:
-    file input_files from featureCounts_to_merge.collect()
+      input:
+      file bam_gene_counter
+      file gtf from gtf_gene_counter.collect()
+      file biotypes_header from ch_biotypes_header.collect()
 
-    output:
-    file 'merged_gene_counts.txt'
+      output:
+      file "${bam_htseqcount.baseName}_gene.htseq-count.txt" into geneCounts, htseqcount_to_merge
+      file "${bam_htseqcount.baseName}_biotype.htseq-count.txt" into gene_counter_logs
+      file "${bam_htseqcount.baseName}_biotype_counts*mqc.{txt,tsv}" into gene_counter_biotype
 
-    script:
-    //if we only have 1 file, just use cat and pipe output to csvtk. Else join all files first, and then remove unwanted column names.
-    def single = input_files instanceof Path ? 1 : input_files.size()
-    def merge = (single == 1) ? 'cat' : 'csvtk join -t -f "Geneid,Start,Length,End,Chr,Strand,gene_name"'
-    """
-    $merge $input_files | csvtk cut -t -f "-Start,-Chr,-End,-Length,-Strand" | sed 's/Aligned.sortedByCoord.out.markDups.bam//g' > merged_gene_counts.txt
-    """
+      script:
+      def strandedness = "no"
+      if (forward_stranded && !unstranded) {
+          htseqcount_direction = "yes"
+      } else if (reverse_stranded && !unstranded){
+          htseqcount_direction = "reverse"
+      }
+      """
+      htseq-count --order pos \
+        --stranded ${strandedness} \
+        --idattr gene_id \
+        --additional-attr gene_name \
+        --mode union \
+        --nonunique all \
+        --format bam \
+        ${bam_htseqcount} \
+        ${gtf} \
+        > ${bam_htseqcount.baseName}_gene.htseq-count.txt
+      # NOTE: ENSEMBL uses "gene_biotype" while vs GENCODE uses "gene_type"
+      htseq-count --order pos \
+        --stranded ${strandedness} \
+        --idattr ${params.fcGroupFeaturesType} \
+        --mode union \
+        --nonunique all \
+        --format bam \
+        ${bam_htseqcount} \
+        ${gtf} \
+        > ${bam_htseqcount.baseName}_biotype.htseq-count.txt
+
+      # Remove lines specifying no alignment, aka starting with two underscores
+      grep -v '^__' ${bam_htseqcount.baseName}_biotype.htseq-count.txt | cat $biotypes_header - >> ${bam_htseqcount.baseName}_biotype_counts_mqc.txt
+      """
+  }
+
+  /*
+   * STEP 9 - Merge htseqcount
+   */
+  process merge_htseqcount {
+      tag "${input_files[0].baseName - '.sorted'}"
+      publishDir "${params.outdir}/gene_counter", mode: 'copy'
+
+      input:
+      file input_files from htseqcount_to_merge.collect()
+
+      output:
+      file 'merged_gene_counts.csv'
+
+      script:
+      //if we only have 1 file, just use cat and pipe output to csvtk. Else join all files first, and then remove unwanted column names.
```

Comment:
I guess we could merge this into a single process with the aforementioned above and depending on which counting tool was used, the handling is (a bit) different. That would reduce the code to a single process instead of two very similar ones :-)

Reply:
Yeah, that would be great 👍 Its just about having less duplicate code in the workflow, as the "main" process looks relatively similar in terms of outputs/inputs and only the actual handling part is quite different, so such a conditional would be best I think 👍